src/test/java/io/jenkins/plugins/sshbuildagents/ssh/agents/AgentConnectionBaseTest.java

```diff
@@ -0,0 +1,166 @@
+import hudson.slaves.DumbSlave;
+import io.jenkins.plugins.sshbuildagents.ssh.mina.SSHApacheMinaLauncher;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.io.IOUtils;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.api.condition.DisabledOnOs;
+import org.junit.jupiter.api.condition.OS;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.junit.jupiter.WithJenkins;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.images.builder.ImageFromDockerfile;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+/**
+ * Base class to test connections to a remote SSH Agent
+ *
+ */
+@Timeout(value = 10, unit = TimeUnit.MINUTES)
+@WithJenkins
+@Testcontainers(disabledWithoutDocker = true)
+@DisabledOnOs(OS.WINDOWS)
+public abstract class AgentConnectionBaseTest {
+
+    public static final String USER = "jenkins";
+    public static final String PASSWORD = "password";
+    public static final String AGENT_WORK_DIR = "/home/jenkins";
+    public static final int SSH_PORT = 22;
+    public static final String SSH_SSHD_CONFIG = "ssh/sshd_config";
+    public static final String DOCKERFILE = "Dockerfile";
+    public static final String SSH_AUTHORIZED_KEYS = "ssh/authorized_keys";
+    public static final String AGENTS_RESOURCES_PATH = "/io/jenkins/plugins/sshbuildagents/ssh/agents/";
+    public static final String LOGGING_PROPERTIES = "remoting_logger.properties";
+
+    protected JenkinsRule j;
+
+    @BeforeEach
+    void beforeEach(JenkinsRule j) {
+        this.j = j;
+        this.j.timeout = 0;
+    }
+
+    @Test
+    void connectionTests() throws IOException, InterruptedException, Descriptor.FormException {
+        Node node = createPermanentAgent(
+                getAgentName(),
+                getAgentContainer().getHost(),
+                getAgentContainer().getMappedPort(SSH_PORT),
+                getAgentSshKeyPath(),
+                getAgentSshKeyPassphrase());
+        waitForAgentConnected(node);
+        assertTrue(isSuccessfullyConnected(node));
+    }
+
+    protected abstract String getAgentName();
+
+    protected abstract GenericContainer<?> getAgentContainer();
+
+    protected String getAgentSshKeyPath() {
+        return null;
+    }
+
+    protected String getAgentSshKeyPassphrase() {
+        return "";
+    }
+
+    protected static boolean isSuccessfullyConnected(Node node) throws IOException, InterruptedException {
+        boolean ret = false;
+        int count = 0;
+        while (count < 30) {
+            Thread.sleep(1000);
+            String log = node.toComputer().getLog();
+            ret = log.contains("Agent successfully connected and online");
+            count++;
+        }
+        return ret;
+    }
```

Comment:
did you mean to just waste time spinning here up to 30 seconds if the agent connects after 1 second?

```suggestion
    protected static boolean isSuccessfullyConnected(Node node) throws IOException, InterruptedException {
        int count = 0;
        while (count < 30) {
            Thread.sleep(1000);
            String log = node.toComputer().getLog();
            if (log.contains("Agent successfully connected and online")) {
              return true;
            }
            count++;
        }
        return false;
    }
```

Reply:
True, I did not check ret to exit from the loop